tests/conftest.py

```diff
@@ -33,82 +28,67 @@
 ]
 
 
-def pytest_addoption(parser: _pytest.config.argparsing.Parser) -> None:
-    parser.addoption(
-        "--test_duration",
-        default="short",
-        choices=["short", "long", "all"],
-        help=
-        "Whether to run short tests (the default), long tests, or all tests. A test is considered short if its timeout is "
-        "less than that specified in the config.")
-
-
-def _get_test_world_size(item: pytest.Item) -> int:
-    world_size_marks = tuple(item.iter_markers(name="world_size"))
-    assert len(world_size_marks) <= 1
-    if len(world_size_marks) == 1:
-        world_size_mark = world_size_marks[0]
-        return world_size_mark.args[0]
-    else:
-        return 1
-
-
-def _get_timeout(item: pytest.Item) -> Optional[float]:
-    # see if it has a timeout marker
-    timeout_mark = item.get_closest_marker("timeout")
-    if timeout_mark is None:
-        return None
-    return timeout_mark.args[0]
-
-
-def _filter_items_for_world_size(items: List[pytest.Item]) -> None:
-    items_to_remove: List[pytest.Item] = []
-    env_world_size = int(os.environ.get('WORLD_SIZE', '1'))
+def pytest_addoption(parser: pytest.Parser) -> None:
+    parser.addoption("--duration",
+                     default="short",
+                     choices=["short", "long", "all"],
+                     help="""Duration of tests, one of short, long, or all.
+                             Tests are short if their timeout < 2 seconds
+                             (configurable threshold). Default: short.""")
+    parser.addoption("--world-size",
+                     default=int(os.environ.get('WORLD_SIZE', 1)),
+                     type=int,
+                     choices=WORLD_SIZE_OPTIONS,
+                     help="""Number of devices. Filters the tests based on their
+                            requested world size. Defaults to 1, and can also
+                            be set by the WORLD_SIZE environment variable.""")
+
+
+def _get_timeout(item: pytest.Item):
+    """Returns the timeout of a test, defaults to 0."""
+    _default = pytest.mark.timeout(0).mark
+    return item.get_closest_marker("timeout", default=_default).args[0]
+
+
+def _get_world_size(item: pytest.Item):
+    """Returns the world_size of a test, defaults to 1."""
+    _default = pytest.mark.world_size(1).mark
+    return item.get_closest_marker("world_size", default=_default).args[0]
+
+
+def pytest_collection_modifyitems(config: pytest.Config, items: List[pytest.Item]) -> None:
+    """Filter tests by world_size (for multi-GPU tests) and duration (short, long, or all)"""
+    timeout_threshold = getattr(config, "_env_timeout", 2.0)
+    duration = config.getoption("duration")
+    world_size = config.getoption("world_size")
+
+    conditions = [
+        lambda item: _get_world_size(item) == world_size,
+    ]
+
+    # separate tests by whether timeout is < or > threshold.
+    if duration == 'short':
```

Comment:
Can you validate that the `duration` is one of `short`, `long`, or `all`? I don't think the argparse validation would work if one were to set the duration int a config file.